contracts/swap/src/queries.rs

```diff
@@ -162,140 +234,166 @@ fn estimate_execution_buy(
-        &querier
-            .query_spot_market_orderbook(
-                &market.market_id,
-                OrderSide::Buy,
-                Some(amount_coin.amount),
-                None,
-            )?
-            .buys_price_level,
-        amount_coin.amount,
-        |l| l.q,
+
+    let top_orders =
+        get_minimum_liquidity_levels(deps, &orders.buys_price_level, amount_coin.amount, |l| l.q)?;
+
+    let average_price = get_average_price_from_orders(&top_orders);
+    let expected_exchange_quantity = amount_coin.amount * average_price;
+    let fee_estimate = expected_exchange_quantity * fee_percent;
+
+    let expected_quantity = expected_exchange_quantity - fee_estimate;
+
+    let worst_price = get_worst_price_from_orders(&top_orders);
+
+    Ok(StepExecutionEstimate {
+        worst_price,
+        result_quantity: expected_quantity,
+        result_denom: counter_denom(market, &amount_coin.denom)?.to_string(),
+        is_buy_order: false,
+        fee_estimate: Some(FPCoin {
+            denom: market.quote_denom.clone(),
+            amount: fee_estimate,
+        }),
+    })
+}
+
+fn estimate_execution_sell_from_target(
+    deps: &Deps<InjectiveQueryWrapper>,
+    querier: &InjectiveQuerier,
+    market: &SpotMarket,
+    amount_coin: FPCoin,
+    fee_percent: FPDecimal,
+) -> StdResult<StepExecutionEstimate> {
+    let available_funds = amount_coin.amount / (FPDecimal::one() + fee_percent); // keep reserve for fee
+    let fee_estimate = amount_coin.amount - available_funds;
+
+    let orders = querier.query_spot_market_orderbook(
+        &market.market_id,
+        OrderSide::Buy,
+        None,
+        Some(available_funds),
     )?;
-    let avg_price = avg_price(&top_orders);
-    let expected_exchange_quantity = amount_coin.amount * avg_price;
-    let expected_fee = expected_exchange_quantity * fee_percent;
-    let expected_quantity = expected_exchange_quantity - expected_fee;
-    deps.api.debug(&format!("Sell exchange: {expected_exchange_quantity}, fee: {expected_fee}, total: {expected_quantity}"));
-    let worst_price = worst_price(&top_orders);
+    let top_orders =
+        get_minimum_liquidity_levels(deps, &orders.buys_price_level, available_funds, |l| {
+            l.q * l.p
+        })?;
+
+    let average_price = get_average_price_from_orders(&top_orders);
+    let expected_input_quantity = amount_coin.amount / average_price;
+    let worst_price = get_worst_price_from_orders(&top_orders);
 
     Ok(StepExecutionEstimate {
         worst_price,
-        result_quantity: round_to_min_tick(expected_quantity, market.min_price_tick_size),
+        result_quantity: round_to_min_tick(expected_input_quantity, market.min_price_tick_size),
         result_denom: counter_denom(market, &amount_coin.denom)?.to_string(),
         is_buy_order: false,
         fee_estimate: Some(FPCoin {
             denom: market.quote_denom.clone(),
-            amount: expected_fee,
+            amount: fee_estimate,
```

Comment:
also, doesn't `estimate_execution_sell_from_target` execute the same logic as buy from source?

Reply:
No, one is using Buy orderbook, one is using Sell orderbook and also one requires the check with `funds_for_margin`. Nonetheless you are right, they are extremely similar. And I noticed one minor difference that I added as TODO comment to double check why it exists.

Reply:
Double checked the difference for the fee, it's indeed correct. Can be explained like this: `estimate_execution_buy_from_source` has a given quote amount, fee is applied to this amount directly as we are doing a buy. `estimate_execution_sell_from_target` on the other hand has a target receipt amount in quote. Fee is not applied to this target amount, but rather it's applied to whatever the swap amount was.